### ISSUE TITLE
Manage IDs for me

### ISSUE THREAD

**Issue by kentcdodds** [edmundhung/conform#108]
With fields, you have to have unique IDs to associate inputs and labels as well as errors (the `aria-errormessage` prop on the input should be assigned to the ID of the error message element if there is an error, and it should be `undefined` if not).

I think conform could definitely manage all this for me. I tried migrating one of my existing forms to conform and I ran into a few things that I think conform should do itself. I added ❗ next to comments that I want to call out specifically:

```tsx
import type { DataFunctionArgs } from '@remix-run/node'
import { json } from '@remix-run/node'
import { Link, useFetcher, useNavigate } from '@remix-run/react'
import { useEffect, useRef, useState } from 'react'
import { AuthorizationError } from 'remix-auth'
import { FormStrategy } from 'remix-auth-form'
import { useForm, parse, conform } from '@conform-to/react'
import { formatError, getFieldsetConstraint } from '@conform-to/zod'
import { z } from 'zod'
import { authenticator } from '~/utils/auth.server'
import { commitSession, getSession } from '~/utils/session.server'
import { passwordSchema, usernameSchema } from '~/utils/user-validation'
import { ErrorList, useFocusInvalid } from '~/utils/forms'

export const LoginFormSchema = z.object({
	username: usernameSchema,
	password: passwordSchema,
	remember: z.boolean().optional(),
})

export async function action({ request }: DataFunctionArgs) {
	const formData = await request.clone().formData()
	const submission = parse(formData)
	const result = LoginFormSchema.safeParse(submission.value)
	if (!result.success) {
		// ❗ it's uncertain to me whether this handles formErrors and fieldErrors
		// or if it's all just flattened...
		submission.error.push(...formatError(result.error))
		return json({ status: 'invalid-form', submission } as const)
	}
	if (submission.type !== 'submit') {
		return json({ status: 'valid-form', submission } as const)
	}

	let userId: string | null = null
	try {
		userId = await authenticator.authenticate(FormStrategy.name, request, {
			throwOnError: true,
		})
	} catch (error) {
		if (error instanceof AuthorizationError) {
			return json(
				{
					status: 'auth-error',
					errors: {
						formErrors: [error.message],
						fieldErrors: {},
					},
				} as const,
				{ status: 400 },
			)
		}
		throw error
	}

	const session = await getSession(request.headers.get('cookie'))
	session.set(authenticator.sessionKey, userId)
	const { remember } = result.data
	const newCookie = await commitSession(session, {
		maxAge: remember
			? 60 * 60 * 24 * 7 // 7 days
			: undefined,
	})
	return json({ status: 'success', errors: null } as const, {
		headers: { 'Set-Cookie': newCookie },
	})
}

export function InlineLogin({ redirectTo }: { redirectTo?: string }) {
	const loginFetcher = useFetcher<typeof action>()
	const navigate = useNavigate()
	const formRef = useRef<HTMLFormElement>(null)
	const [form, { username, password, remember }] = useForm({
		constraint: getFieldsetConstraint(LoginFormSchema),
		state:
			loginFetcher.data?.status === 'invalid-form'
				? loginFetcher.data.submission
				: undefined,
	})

	// ❗ I don't want to have to set these at all. It should be possible to
	// override, but these should just default for me.
	const usernameId = 'username'
	const usernameErrorId = 'username-error'
	const passwordId = 'password'
	const passwordErrorId = 'password-error'
	const rememberId = 'remember'
	const fields = {
		username: {
			labelProps: { htmlFor: usernameId },
			fieldProps: {
				id: usernameId,
				'aria-errormessage': username.error ? usernameErrorId : undefined,
				...conform.input(username.config),
			},
			// ❗ I'd really prefer it to be possible to have an array of errors
			errorUI: username.error ? (
				<ErrorList errors={[username.error]} id={usernameErrorId} />
			) : null,
		},
		password: {
			labelProps: { htmlFor: passwordId },
			fieldProps: {
				id: passwordId,
				'aria-errormessage': password.error ? passwordErrorId : undefined,
				...conform.input(password.config),
			},
			errorUI: password.error ? (
				<ErrorList errors={[password.error]} id={passwordErrorId} />
			) : null,
		},
		remember: {
			labelProps: { htmlFor: rememberId },
			fieldProps: {
				id: rememberId,
				...conform.input(remember.config),
			},
			errorUI: null,
		},
	} as const

	const formErrorUI = form.error ? (
		<ErrorList errors={[form.error]} id="form-error" />
	) : null

	const success = loginFetcher.data?.status === 'success'
	useEffect(() => {
		if (!redirectTo) return
		if (!success) return

		navigate(redirectTo)
	}, [success, redirectTo])

	// ❗ doing this to be able to control the noValidate prop
	const [hydrated, setHydrated] = useState(false)
	useEffect(() => {
		setHydrated(true)
	}, [])

	// ❗ this would be really nice to have built-in
	useFocusInvalid(formRef.current, {
		formErrors: [form.error],
		fieldErrors: {
			username: [username.error],
			password: [password.error],
		},
	})

	return (
		<div>
			<div className="mx-auto w-full max-w-md px-8">
				<loginFetcher.Form
					method="post"
					action="/resources/login"
					className="space-y-6"
					// ❗ I have to manually set the aria-errormessage here
					aria-errormessage={formErrorUI ? 'form-error' : undefined}
					{...form.props}
					// ❗ I have to override the ref 😬 I should be able to provide my own
					// ref without messing up conform...
					ref={formRef}
					// ❗ I'm passing noValidate here to make sure noValidate only gets set
					// to true once we've successfully hydrated
					noValidate={hydrated}
				>
					<div>
						<label
							className="block text-sm font-medium text-gray-700"
							{...fields.username.labelProps}
						>
							Username
						</label>
						<div className="mt-1">
							<input
								autoComplete="username"
								className="w-full rounded border border-gray-500 px-2 py-1 text-lg"
								{...fields.username.fieldProps}
							/>
							{/*
								❗ notice the lack of logic here. I just render it and the logic
								above ensures that the errorUI is `null` if there is no error
							*/}
							{fields.username.errorUI}
						</div>
					</div>

					<div>
						<label
							className="block text-sm font-medium text-gray-700"
							{...fields.password.labelProps}
						>
							Password
						</label>
						<div className="mt-1">
							<input
								autoComplete="current-password"
								className="w-full rounded border border-gray-500 px-2 py-1 text-lg"
								{...fields.password.fieldProps}
								type="password"
							/>
							{fields.password.errorUI}
						</div>
					</div>

					<div className="flex items-center">
						<input
							className="h-4 w-4 rounded border-gray-300 text-blue-600 focus:ring-blue-500"
							{...fields.remember.fieldProps}
							// ❗ I have to manually set the type here. I'd love to either have a
							// conform.checkbox or have conform.input be able to handle this automatically
							type="checkbox"
						/>
						<label
							className="ml-2 block text-sm text-gray-900"
							{...fields.remember.labelProps}
						>
							Remember me
						</label>
					</div>

					{formErrorUI}

					<div className="flex items-center justify-between gap-6">
						<button
							type="submit"
							className="w-full rounded bg-blue-500  py-2 px-4 text-white hover:bg-blue-600 focus:bg-blue-400"
						>
							Log in
						</button>
					</div>
				</loginFetcher.Form>
				<div className="flex justify-around pt-6">
					<Link to="/signup" className="text-blue-600 underline">
						New here?
					</Link>
					<Link to="/forgot-password" className="text-blue-600 underline">
						Forgot password?
					</Link>
				</div>
			</div>
		</div>
	)
}
```

There are several things in this code example that I think could use some cleanup and help from the conform library. My `fields` object I think has things that conform could do for me.

It's possible conform *does* handle some or all of this stuff and I just missed it in the docs. Any direction is welcome.

**Comment by kentcdodds** [comment]
> Conform support generating these IDs

Ah! That makes sense. I assumed that since I had to pass a prop it wouldn't apply the IDs to everything, but support for React 17 necessitating me passing this makes sense.

> You can achieve this by setting the fallbackNative options on [useForm](https://conform.guide/api/react#useform) hook to true. (The name of the option is properly confusing, suggestion is welcome!)

Ah, yeah, that makes sense. I think `fallbackNative` is a fine name 👍 My bad for not reading the docs closely enough.

> `useFocusInvalid`

Actually, this is my bad, I already had this hook lying around so I just used it and assumed it wouldn't work. I think moving focus is probably the right call, but I'm not 100% certain 😅

> the format `${formId}-error`. Will it work for you?

That works great 👍

> But if you really need to access the ref.

I always like to make it possible for people to pass a ref themselves, and in general I've had more success when the ref that's used is the one created by the developer, not the library. So what you could do is allow me to pass the `ref` to `useForm` and then the `useForm` hook can use the provided ref and just fallback to its own if none is provided.

> ❗ notice the lack of logic here. I just render it and the logic above ensures that the errorUI is `null` if there is no error

How about something like this:

```tsx
const [form, {username}] = useForm({
  errorComponent: ErrorList
})

return <div>{username.errorUI}</div>

// ...

function ErrorList({id, errors}) {
  return <ul id={id}>{errors.map(...)}</ul>
}
```

Alternatively it could be a "render prop"-like API so people don't have to make a component for it:

```tsx
const [form, {username}] = useForm({
  errorUI: ({id, errors}) => <ul id={id}>{errors.map(...)}</ul>
})
```

I don't mind either way. I'm guessing the component approach would be more acceptable by more people.

> But I think what you wanna do ultimately is to have it derived from `getFieldsetConstraint(LoginFormSchema)`, right?

This is what I do. Pretty much if it's a boolean schema type then I set the `type` prop to `checkbox`.

**Comment by edmundhung** [comment]
Conform v0.6 is out now with multiple error and form level aria-attributes support. More details can be found on the [release note](https://github.com/edmundhung/conform/releases/tag/v0.6.0).

> So what you could do is allow me to pass the ref to useForm and then the useForm hook can use the provided ref and just fallback to its own if none is provided.

Make sense 👍🏼 This should be easy to add quickly.

> Alternatively it could be a "render prop"-like API so people don't have to make a component for it:

```tsx
const [form, {username}] = useForm({
  errorUI: ({id, errors}) => <ul id={id}>{errors.map(...)}</ul>
})
```

How about something like this?

```tsx
const [form, { username }] = useForm({
  // This basically enable you to transform the field config
  // structure to any structure you like 
  transformField: (config) => {
    return {
      labelProps: { htmlFor: config.id },
      inputProps: { id: config.id, name: config.name },
      errorUI: <ul id={config.errorId}>{config.errors.map(...)}</ul>,
    };
  },
})

// username.labelProps
// username.inputProps
// username.errorUI
```

One concern I see is this breaks the ability to pass a field to another conform hook directly as the data structure is changed. You will also need to pass the same transform function to every hook calls for nested / list, unless we do this (?)

```tsx
const [form, { options }] = useForm({
  transformField: (config) => {
    return {
      labelProps: { htmlFor: config.id },
      inputProps: { id: config.id, name: config.name },
      errorUI: <ul id={config.errorId}>{config.errors.map(...)}</ul>,

      // Keep the config structure so you can pass it to other hooks
      config: {
        ...config,
        // So you can bring the transform function to another hook as well.. (?)
        transformField: this,
      },
    };
  },
})

// So now you can do this
const optionList = useFieldList(form.ref, options.config)
```

**Comment by kentcdodds** [comment]
Ah yeah, I don't think I like the idea of breaking composability. Maybe instead it's simply a component that I can pass the config to?

**Comment by lifeiscontent** [comment]
@edmundhung it seems that for things like checkboxes/radios the ids that are used are not unique which can cause issues for the error integrations, focus input when clicking labels

**Comment by aaronadamsCA** [comment]
Hi! I am new to this library and I love it so far. I just wanted to voice my support for always generating field IDs. This has been my biggest surprise so far.

Currently you can avoid field ID collisions by passing a unique form ID. This is a good feature! But I think it should be optional; if you don't pass a form ID, I think the field IDs should be defined anyway, without a form ID prefix. This should really be safe enough for most cases, and would save our implementation many runtime null checks.

**Comment by edmundhung** [comment]
We will be generating a default id with the useId() hook in the next version. But it means we will only support react 18 or above going forward.

**Comment by MattyBalaam** [comment]
Where are we with this at the moment, we seem to get a prop `aria-describedby` passed down, but not `aria-errormessage`.

Is this correct?

**Comment by edmundhung** [comment]
Correct. It's because [aria-errormessage is not well supported as aria-describedby](https://www.digitala11y.com/aria-errormessageproperties)

**Comment by MattyBalaam** [comment]
Ineresting, I had assumed support was better than it is, maybe one day then.
